Added PDF-Page-Chunk-Extractor for image extraction (Issue 703)

kamir opened the pull request.

aaron-siegel reviewed (changes requested):
Thanks for the contribution! Can you please add a description to the PR summarizing the changes?

kamir commented:
I changed test scripts to use pytest and added the initial PR description.

aaron-siegel commented on pixeltable/catalog/catalog.py:
```diff
@@ -930,9 +930,9 @@ def get_table(self, path: Path) -> Table:
 
     @_retry_loop(for_write=True)
-    def drop_table(self, path: Path, if_not_exists: IfNotExistsParam, force: bool) -> None:
+    def drop_table(self, path: str, if_not_exists: IfNotExistsParam, force: bool) -> None:
```
This change shouldn't be necessary and might break other features.

kamir replied:
Reverted the change.

aaron-siegel commented on pixeltable/env.py:
```diff
@@ -309,10 +309,15 @@ def _log_filter(self, record: logging.LogRecord) -> bool:
-            for module_name in path_parts[:max_idx]:
-                if module_name in self._module_log_level and record.levelno >= self._module_log_level[module_name]:
-                    return True
+            try:
```
Can this be recast as a conditional rather than an exception handler?

`if 'pixeltable' in path_parts:`

If that works it'd be cleaner.

kamir replied:
reverted this change.

aaron-siegel commented on pixeltable/iterators/pdf_page_extractor.py:
```diff
@@ -0,0 +1,60 @@
+
+    @classmethod
+    def input_schema(cls) -> dict[str, ColumnType]:
+        return {'document': StringType(nullable=False)}
```
input should be `DocumentType` not `StringType`

aaron-siegel commented on tests/iterators/test_pdf_page_extractor.py:
```diff
@@ -0,0 +1,62 @@
+class TestPdfPageExtractor:
+    @pytest.mark.parametrize("limit", [2])  # Only test a few PDFs to keep runtime low
+    def test_pdf_page_extraction(self, limit: int) -> None:
+        pdfs = find_pdfs("tests/data/documents", limit=limit)
```
The `find_pdfs` function shouldn't be necessary; just use

`glob.glob('tests/data/documents/**/*.pdf')`

Better yet, use the existing `utils.get_documents()` which uses more robust path resolution, and filter on pdfs:

`pdfs = [path for path in get_documents() if path.endswith('.pdf')]`

aaron-siegel commented on tests/iterators/test_pdf_page_extractor.py:
```diff
@@ -0,0 +1,62 @@
+
+                reopened_img = Image.open(io.BytesIO(img_bytes))
+                assert reopened_img.size == chunk["image"].size
+                break  # Only test the first page to keep the test fast
```
That would seem to cripple the test since it fails to validate the actual iteration behavior.

Maybe limit to the first 3 pages with something like this?

`for chunk in itertools.islice(extractor, 3):`

kamir replied:
Code changed

aaron-siegel commented on tests/test_pdf_chunks.py:
```diff
@@ -0,0 +1,71 @@
+from pixeltable.iterators.pdf_page_extractor import PdfPageExtractor  # adjust path as needed
+from pixeltable.catalog import Catalog
+
+class TestPdfExtraction:
```
What is this test doing that's different from the other one? Would just one of them suffice and/or can they be condsolidated into one test class?

aaron-siegel reviewed (changes requested):
You will want to run `make format` to get the code formatting right, and then `make check` to run the static checks against your changes, in order to ensure that our CI accepts the PR.

kamir commented on pixeltable/iterators/pdf_page_extractor.py:
```diff
@@ -0,0 +1,60 @@
+        clean_text = text.replace('\x00', '')
+        
+        # Render image
+        pix = page.get_pixmap(dpi=300)
```
Done.

kamir commented on pixeltable/iterators/pdf_page_extractor.py:
```diff
@@ -0,0 +1,60 @@
+    text: str
+    page: int
+    image_bytes: bytes
+    image: Image
```
Update: Image => PIL.Image.Image

kamir commented on tests/conftest.py:
```diff
@@ -55,7 +55,12 @@ def pxt_test_harness() -> Iterator[None]:
 @pytest.fixture(scope='session')
-def init_env(tmp_path_factory: pytest.TempPathFactory, worker_id: int) -> None:
+def init_env(tmp_path_factory: pytest.TempPathFactory, request: pytest.FixtureRequest) -> None:
+    # worker_id is provided by pytest-xdist. If not running in parallel, we use a default.
```
Reverted this change.

kamir replied:
change reverted

kamir commented on tests/test_pdf_chunks.py:
```diff
@@ -0,0 +1,71 @@
+        Catalog.get().drop_table('pdf_docs', if_not_exists=True, force=True)
+
+        # Create base table
+        docs = pxt.create_table('pdf_docs', {'doc': pxt.String}, if_exists='replace_force')
```
fixed it.

kamir commented on tests/test_pdf_chunks.py:
```diff
@@ -0,0 +1,71 @@
+        docs = pxt.create_table('pdf_docs', {'doc': pxt.String}, if_exists='replace_force')
+
+        # Insert documents
+        insert_result = docs.insert([{'doc': p} for p in pdf_paths])
```
Updated the code.

kamir commented on pixeltable/iterators/pdf_page_extractor.py:
```diff
@@ -0,0 +1,60 @@
+    image_bytes: bytes
+    image: Image
+
+class PdfPageExtractor(ComponentIterator):
```
Good point. I considered implementing the new feature into the DocumentSplitter, but kept it separate to keep PDF-specific logic (like page rendering and image handling) cleanly isolated. 
My main goal is separation of concerns — but I’ve aligned the interface and schema to stay compatible with the existing pattern. 
Next step on my agenda for PDF file processing is diving into tables, figures, citations and semantic chunking. I guess this kind of logic could disturb the pure DocumentSplitter.
Happy to revisit if we want tighter integration.

kamir commented:
I updated the PR and addressed all comments.
The comments to minor changes are closed, and the main question is commented.

The tests passed localy.

aaron-siegel commented on Makefile.MK:
```diff
@@ -0,0 +1,32 @@
+PROJECT_DIR := $(shell pwd)
```
Why are we adding an additional makefile to the repo?

aaron-siegel commented on pixeltable/catalog/catalog.py:
```diff
@@ -932,7 +932,7 @@ def get_table(self, path: Path) -> Table:
     def drop_table(self, path: Path, if_not_exists: IfNotExistsParam, force: bool) -> None:
         tbl = self._get_schema_object(
-            path,
+            Path(path),
```
Spurious

aaron-siegel commented on pyproject.toml:
```diff
@@ -155,6 +155,7 @@ ollama = ">=0.5.1"
 llama-cpp-python = { version = ">=0.3.9", platform = "!=win32" }  # Windows build of llama_cpp has issues in CI
 mcp = ">=1.9.4"
 pixeltable-yolox = "==0.4.2"
+flake8 = "^7.3.0"
```
Not necessary - we already use `ruff` which subsumes `flake8`.

kamir replied:
Removed flake8 = "^7.3.0"

aaron-siegel commented on tests/iterators/test_pdf_page_extractor.py:
```diff
@@ -0,0 +1,62 @@
+
+                reopened_img = Image.open(io.BytesIO(img_bytes))
+                assert reopened_img.size == chunk['image'].size
+                break  # Only test the first page to keep the test fast
```
Shouldn't the `break` be removed now that the iteration is bounded? Otherwise adding the `islice` has no effect?

kamir replied:
Removed break.

aaron-siegel commented on tests/iterators/test_pdf_page_extractor.py:
```diff
@@ -0,0 +1,62 @@
+
+
+class TestPdfPageExtractor:
+    @pytest.mark.parametrize('limit', [2])  # Only test a few PDFs to keep runtime low
```
No need for `parametrize` unless there are actually multiple variants being tested.
